### ISSUE TITLE
Last selected LLM provider is not persisted anymore

### ISSUE THREAD

**Issue by mydeveloperplanet** [devoxx/DevoxxGenieIDEAPlugin#181]
Using DevoxxGenie 0.2.5

1. Select LMStudio
2. Switch to another project or open a new IntelliJ window
3. The LLM provider Ollama is selected. Before, the last selected LLM provider was persisted

It also occurs when I have selected GPT4All. It seems that always Ollama is selected when starting an IntelliJ instance.

**Comment by stephanj** [comment]
Can you first upgrade to v0.2.8

**Comment by mydeveloperplanet** [comment]
Sure, but now Exo seems to be selected all the time

**Comment by stephanj** [comment]
The new version sorts them alphabetically, that's why exo is shown first.
Can you give me a step-by-step how to reproduce it, because I'm always seeing my latest selected LLM provider.

**Comment by mydeveloperplanet** [comment]
How to reproduce:
1. Start LMStudio
2. Start IntelliJ
3. Select LMStudio in DevoxxGenie
4. use DevoxxGenie to explain something (but I do not think is actually needed)
5. Switch to another project, same window (open the project in a new window has no influence)
6. Open DevoxxGenie if necessary: Exo is selected
7. Switch to the previous project, same window
8. Exo is selected

I also tried to invalidate all IntelliJ caches, but this did not resolve the problem.

I also shut down IntelliJ and removed the ~/.config/JetBrains/IdeaIC2024.1/options/DevoxxGenieSettingsPlugin.xml
Executed the same steps.
I do see that after step 5, the selectedProvider is LMStudio in the settings file. That is correct. So the problem probably occurs when reading the settings file.

**Comment by stephanj** [comment]
Can you give this version a try, this will store the LLM provider per project. 
However the LLM model is (not yet) restored correctly only the provider.

[DevoxxGenie-0.2.9.zip](https://github.com/user-attachments/files/16410055/DevoxxGenie-0.2.9.zip)

**Comment by mydeveloperplanet** [comment]
Did not change anything, I think I will need to debug this and see whether I can see an error somewhere. I will try to do so one of the coming days

**Comment by stephanj** [comment]
BTW You can check what's being stored (per project) by going to the following directory (assuming you're on a Mac)

```shell
$ cd /Users/{{{user}}}/Library/Application Support/JetBrains/IntelliJIdea2024.1/options
$ more DevoxxGenieSettingsPlugin.xml
```

![image](https://github.com/user-attachments/assets/e3971b72-f520-4c8f-bead-2d2a13238e24)

**Comment by mydeveloperplanet** [comment]
As far as I can see, the correct LLM provider is stored.
When I debug, I notice that DevoxxGenieToolWindowContent.setupUI() calls setLastSelectedProvider when I open a project. But here, it already contains Exo as value.
If you have any hints where I can debug any further, please let me know.

**Comment by stephanj** [comment]
⚠️ Claude Sonnet 3.5: 

I'll explain the flow of how the last selected Language Model (LLM) and model name are handled in the DevoxxGenie IntelliJ plugin. This process involves several components working together to maintain and use the selected LLM provider and model name. Let's break it down step by step with code examples.

1. Storing the Last Selected Provider and Model

The `DevoxxGenieStateService` class is responsible for storing the plugin's state, including the last selected provider and model name. Here are the relevant fields:

```java
@Getter
@Setter
@State(
    name = "com.devoxx.genie.ui.SettingsState",
    storages = @Storage("DevoxxGenieSettingsPlugin.xml")
)
public final class DevoxxGenieStateService implements PersistentStateComponent<DevoxxGenieStateService> {
    // ...
    private String selectedProvider;
    private String selectedLanguageModel;
    // ...
}
```

2. Updating the Selected Provider and Model

When a user selects a new provider or model, the `DevoxxGenieToolWindowContent` class updates these values in the `DevoxxGenieStateService`. This happens in the `handleModelProviderSelectionChange` and `processModelNameSelection` methods:

```java
private void handleModelProviderSelectionChange(@NotNull ActionEvent e) {
    // ...
    ModelProvider modelProvider = (ModelProvider) comboBox.getSelectedItem();
    if (modelProvider != null) {
        DevoxxGenieStateService.getInstance().setSelectedProvider(modelProvider.getName());
        updateModelNamesComboBox(modelProvider.getName());
        // ...
    }
    // ...
}

private void processModelNameSelection(@NotNull ActionEvent e) {
    if (e.getActionCommand().equals(Constant.COMBO_BOX_CHANGED)) {
        LanguageModel selectedModel = (LanguageModel) modelNameComboBox.getSelectedItem();
        if (selectedModel != null) {
            DevoxxGenieStateService.getInstance().setSelectedLanguageModel(selectedModel.getModelName());
        }
    }
}
```

3. Using the Last Selected Provider and Model

When creating a new chat context, the `ChatMessageContextUtil` class uses the last selected provider and model name to create the `ChatMessageContext`:

```java
public static @NotNull ChatMessageContext createContext(Project project,
                                                        String userPromptText,
                                                        LanguageModel languageModel,
                                                        ChatModelProvider chatModelProvider,
                                                        @NotNull DevoxxGenieStateService stateService,
                                                        @NotNull String actionCommand,
                                                        EditorFileButtonManager editorFileButtonManager,
                                                        String projectContext,
                                                        boolean isProjectContextAdded,
                                                        int totalFileCount) {
    ChatMessageContext context = ChatMessageContext.builder()
        .project(project)
        .name(String.valueOf(System.currentTimeMillis()))
        .userPrompt(userPromptText)
        .userMessage(UserMessage.userMessage(userPromptText))
        .languageModel(languageModel)
        // ...
        .build();

    // ...

    return context;
}
```

4. Creating the Chat Model

The `ChatModelProvider` class uses the selected provider and model name to create the appropriate `ChatLanguageModel` or `StreamingChatLanguageModel`:

```java
public class ChatModelProvider {
    // ...
    public ChatLanguageModel getChatLanguageModel(@NotNull ChatMessageContext chatMessageContext) {
        ChatModel chatModel = initChatModel(chatMessageContext);
        return getFactory(chatMessageContext).createChatModel(chatModel);
    }

    public StreamingChatLanguageModel getStreamingChatLanguageModel(@NotNull ChatMessageContext chatMessageContext) {
        ChatModel chatModel = initChatModel(chatMessageContext);
        return getFactory(chatMessageContext).createStreamingChatModel(chatModel);
    }

    private @NotNull ChatModelFactory getFactory(@NotNull ChatMessageContext chatMessageContext) {
        ModelProvider provider = Optional.ofNullable(chatMessageContext.getLanguageModel())
            .map(LanguageModel::getProvider)
            .orElse(DEFAULT_PROVIDER);

        ChatModelFactory factory = factories.get(provider);
        if (factory == null) {
            throw new IllegalArgumentException("No factory for provider: " + provider);
        }
        return factory;
    }
    // ...
}
```

5. Executing the Prompt

Finally, when executing a prompt, the `ChatPromptExecutor` class uses the created `ChatLanguageModel` or `StreamingChatLanguageModel` to generate a response:

```java
public class ChatPromptExecutor {
    // ...
    public void executePrompt(@NotNull ChatMessageContext chatMessageContext,
                              PromptOutputPanel promptOutputPanel,
                              Runnable enableButtons) {
        isRunning = true;
        new Task.Backgroundable(chatMessageContext.getProject(), "Working...", true) {
            @Override
            public void run(@NotNull ProgressIndicator progressIndicator) {
                if (chatMessageContext.isWebSearchRequested()) {
                    new WebSearchExecutor().execute(chatMessageContext, promptOutputPanel, () -> {
                        isRunning = false;
                        enableButtons.run();
                    });
                } else if (DevoxxGenieStateService.getInstance().getStreamMode()) {
                    streamingPromptExecutor.execute(chatMessageContext, promptOutputPanel, () -> {
                        isRunning = false;
                        enableButtons.run();
                    });
                } else {
                    nonStreamingPromptExecutor.execute(chatMessageContext, promptOutputPanel, () -> {
                        isRunning = false;
                        enableButtons.run();
                    });
                }
            }
        }.queue();
    }
    // ...
}
```

In summary, the flow of the last selected LLM and model name is as follows:

1. The user selects a provider and model in the UI.
2. The `DevoxxGenieToolWindowContent` updates the `DevoxxGenieStateService` with the new selections.
3. When creating a new chat context, `ChatMessageContextUtil` uses the stored provider and model name.
4. The `ChatModelProvider` creates the appropriate `ChatLanguageModel` or `StreamingChatLanguageModel` based on the stored selections.
5. The `ChatPromptExecutor` uses the created model to generate responses.

This design allows the plugin to maintain the user's last selected LLM provider and model name across sessions, ensuring a consistent experience and allowing for easy switching between different LLMs and models.

**Comment by mydeveloperplanet** [comment]
Thanks :-)
It looks like a race condition. 
I added a console output in DevoxxGenieStateService.loadState (begin and end and I print the state.getSelectedProvider).
I added a console output in DevoxxGenieToolWindowContent before and after setupUI and I print the modelProvider.getName() in setLastSelectedProvider.

The output is the following:
```
before setupUI
loadState begin LMStudio
loadState end LMStudio
setLastSelectedProvider Exo (Experimental)
after setupUI
setLastSelectedProvider Exo (Experimental)
```
The state is loaded correctly, but at the same time, DevoxxGenieToolWindowContent is started.
I haven't found yet where loading the state leads to setting the selected provider.

**Comment by stephanj** [comment]
Thanks for delving deeper. I've asked Claude to have a look and this is what it produced:

It appears that there's a race condition in restoring the last selected Language Model (LLM) provider. Let's analyze the problem and propose a solution.

The race condition seems to occur because the UI setup and the state restoration are happening concurrently or in an uncontrolled order. To fix this, we need to ensure that the state is fully loaded before setting up the UI and restoring the last selected provider.

Here's a proposed solution:

1. Introduce a flag to track whether the state has been fully loaded.
2. Defer the UI setup and provider selection until the state is loaded.
3. Use a callback mechanism to ensure proper sequencing.

Let's modify the `DevoxxGenieToolWindowContent` class to implement this solution:

```java
public class DevoxxGenieToolWindowContent implements SettingsChangeListener,
                                                     LLMSettingsChangeListener,
                                                     ConversationStarter,
                                                     CustomPromptChangeListener {

    private boolean isStateLoaded = false;
    private String lastSelectedProvider = null;

    // ... other existing fields ...

    public DevoxxGenieToolWindowContent(@NotNull ToolWindow toolWindow) {
        project = toolWindow.getProject();

        // Defer UI setup
        DevoxxGenieStateService.getInstance().addLoadListener(this::onStateLoaded);

        setupMessageBusConnection(toolWindow);
    }

    private void onStateLoaded() {
        isStateLoaded = true;
        lastSelectedProvider = DevoxxGenieStateService.getInstance().getSelectedProvider();
        
        ApplicationManager.getApplication().invokeLater(() -> {
            setupUI();
            restoreLastSelectedProvider();
        });
    }

    private void setupUI() {
        // ... existing UI setup code ...
    }

    private void restoreLastSelectedProvider() {
        if (lastSelectedProvider != null) {
            for (int i = 0; i < modelProviderComboBox.getItemCount(); i++) {
                ModelProvider provider = modelProviderComboBox.getItemAt(i);
                if (provider.getName().equals(lastSelectedProvider)) {
                    modelProviderComboBox.setSelectedIndex(i);
                    updateModelNamesComboBox(lastSelectedProvider);
                    break;
                }
            }
        }
    }

    // ... other existing methods ...
}
```

Now, let's modify the `DevoxxGenieStateService` to support this new approach:

```java
@State(
    name = "com.devoxx.genie.ui.SettingsState",
    storages = @Storage("DevoxxGenieSettingsPlugin.xml")
)
public final class DevoxxGenieStateService implements PersistentStateComponent<DevoxxGenieStateService> {

    private List<Runnable> loadListeners = new ArrayList<>();

    // ... other existing fields and methods ...

    public void addLoadListener(Runnable listener) {
        loadListeners.add(listener);
    }

    @Override
    public void loadState(@NotNull DevoxxGenieStateService state) {
        XmlSerializerUtil.copyBean(state, this);
        initializeDefaultCostsIfEmpty();
        initializeDefaultPrompts();
        
        // Notify all listeners that the state has been loaded
        for (Runnable listener : loadListeners) {
            listener.run();
        }
    }

    // ... other existing methods ...
}
```

These changes accomplish the following:

1. We introduce an `isStateLoaded` flag and a `lastSelectedProvider` field in `DevoxxGenieToolWindowContent`.
2. The UI setup is deferred until the state is loaded by using a callback mechanism.
3. We add a `loadListener` to `DevoxxGenieStateService` that will be called when the state is fully loaded.
4. The `onStateLoaded` method in `DevoxxGenieToolWindowContent` sets up the UI and restores the last selected provider after the state is loaded.
5. We use `ApplicationManager.getApplication().invokeLater()` to ensure that UI updates happen on the EDT (Event Dispatch Thread).

This approach should resolve the race condition by ensuring that the state is fully loaded before any UI setup or provider selection occurs. The sequence will now be:

1. `DevoxxGenieStateService` loads the state.
2. `onStateLoaded` callback is triggered.
3. UI is set up.
4. Last selected provider is restored.

This should eliminate the race condition and ensure consistent behavior when restoring the last selected LLM provider.

**Comment by mydeveloperplanet** [comment]
This introduces a new race condition :-) The `DevoxxGenieStateService.loadState` is finished before `addLoadListener` is invoked. However, it gave me some ideas to create a more robust solution, but that is something for the weekend.

**Comment by stephanj** [comment]
The last selected provider and language model should also be stored per project as follows:

```java
public final class DevoxxGenieStateService... {
    private Map<String, String> lastSelectedProvider;
    private Map<String, String> lastSelectedLanguageModel;

    //... 

    public void setSelectedLanguageModel(@NotNull Project project, String selectedLanguageModel) {
        if (lastSelectedLanguageModel == null) {
            lastSelectedLanguageModel = new HashMap<>();
        }
        lastSelectedLanguageModel.put(project.getLocationHash(), selectedLanguageModel);
    }

    public String getSelectedLanguageModel(@NotNull Project project) {
        if (lastSelectedLanguageModel != null) {
            return lastSelectedLanguageModel.getOrDefault(project.getLocationHash(), "");
        } else {
            return "";
        }
    }

    public void setSelectedProvider(@NotNull Project project, String selectedProvider) {
        if (lastSelectedProvider == null) {
            lastSelectedProvider = new HashMap<>();
        }
        lastSelectedProvider.put(project.getLocationHash(), selectedProvider);
    }

    public String getSelectedProvider(@NotNull Project project) {
        if (lastSelectedProvider != null) {
            return lastSelectedProvider.getOrDefault(project.getLocationHash(), ModelProvider.Ollama.getName());
        } else {
            return ModelProvider.Ollama.getName();
        }
    }
}
```

as a result, the project must be passed to the related methods:

```
DevoxxGenieStateService.getInstance().setSelectedLanguageModel(project, selectedModel.getModelName());
```

This will allow us to have different LLM providers and model names per (open) project.

**Comment by mydeveloperplanet** [comment]
Fixed the initialization of the language model combox when no devoxxgenie settings are present.
Fixed the restoring of the language model.

I would prefer to add storage of the provider and language model per project in a different branch, that is more like a feature while this issue is more a bugfix :-)
And I have some additional questions about this, like:
What to do if a project does not have a persistent record yet? Should the last selected provider and language model you selected be used then?

**Comment by stephanj** [comment]
When no persistent record is set we currently fall back to "Ollama".   
This can always be "improved" based on users feedback  :)

```
public String getSelectedProvider(@NotNull Project project) {
        if (lastSelectedProvider != null) {
            return lastSelectedProvider.getOrDefault(project.getLocationHash(), ModelProvider.Ollama.getName());
        } else {
            return ModelProvider.Ollama.getName();
        }
    }
```
